xds/internal/xdsclient/authority.go

```diff
@@ -416,6 +500,76 @@ func (a *authority) handleADSResourceDoesNotExist(rType xdsresource.Type, resour
 	}
 }
 
+// handleRevertingToPrimaryOnUpdate is called when a resource update is received
+// from a server that is not the current active server. This method ensures that
+// all lower priority servers are closed and the active server is reverted to
+// the highest priority server that has sent an update.
+//
+// This method is only executed in the context of a serializer callback.
+func (a *authority) handleRevertingToPrimaryOnUpdate(serverConfig *bootstrap.ServerConfig) {
+	if a.activeXDSChannel != nil && a.activeXDSChannel.sc.Equal(serverConfig) {
+		// If the resource update is from the current active server, nothing
+		// needs to be done from fallback point of view.
+		return
+	}
+
+	if a.logger.V(2) {
+		a.logger.Infof("Received update from non-active server %q", serverConfig)
+	}
+
+	// If the resource update is not from the current active server, it means
+	// that we have received an update from a higher priority server and we need
+	// to revert back to it. This method guarantees that when an update is
+	// received from a server, all lower priority servers are closed.
+	serverIdx := 0
+	for _, cfg := range a.xdsChannelConfigs {
+		if cfg.sc.Equal(serverConfig) {
+			break
+		}
+		serverIdx++
```

Comment:
Here and the fallback operation above:

You could just read the index and read it into serverIdx right before break if you want, might be cleaner/easier to understand than iterating by 1 every node in the slice you search.